src/path-mapping.ts

```diff
@@ -0,0 +1,131 @@
+import type * as ts from 'typescript';
+import { join as joinPath } from 'path';
+
+// Path mapper returns a list of mapped specifiers or `null` if the
+// given `specifier` was not mapped.
+type PathMapper = (specifier: string) => string[] | null;
+
+export function createPathMapper(
+  compilerOptions: ts.CompilerOptions
+): PathMapper {
+  if (compilerOptions.paths) {
+    if (!compilerOptions.baseUrl) {
+      throw new Error(`Compiler option 'baseUrl' required when 'paths' is set`);
```

Comment:
Is `compilerOptions.baseUrl` guaranteed to be an absolute path, or do we need to resolve it relative to the tsconfig's location?

Reply:
Good point. I assumed that we always get the configuration from [`ts.readConfigFile()` in `src/configuration.ts`](https://github.com/TypeStrong/ts-node/blob/8949654f2b7050b2b692eefe14b225d3f6e7901c/src/configuration.ts#L113) and the function resolves the base URL relative to the config file. But there are probably other ways to set the base URL like through the command line, API, or environment.

To address this I think it make sense to resolve any base URL that is not a URL but a relative path to the current directory.

Reply:
I see, we almost always get the configuration from `ts.readConfigFile()`, so we should be good.  I think the only way to set it otherwise is via the API?  Since we don't support all of `tsc`'s command-line flags.

Maybe the correct approach is to resolve `baseUrl` once within `create()` so that we have a reliable, absolute `baseUrl` to be used elsewhere.

Reply:
> Maybe the correct approach is to resolve baseUrl once within create() so that we have a reliable, absolute baseUrl to be used elsewhere.

Alternatively, we could require `baseUrl` to be absolute and throw an error otherwise. This would allow us to avoid implicit, environment-dependent behavior for users of the API. This might prevent some confusion when path mapping doesn’t work as expected but only fails at the resolution stage instead of the construction stage. (For example a user may think the base URL they set on the API is considered relative to `tsconfig.json` and their scripts work whenever they are run from the project root. But then they run the script from a subdirectory and it fails.)

Reply:
Sounds good, let's do that.  In `create()` we throw an error if baseUrl is not absolute.

Reply:
Actually, scratch that, seems we always pass options through TypeScript's API to be normalized, even when provided via our API:

```
require('ts-node').create({compilerOptions: {baseUrl: "foobar"}}).config.options.baseUrl
```

So I guess we're all good.